ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -584,10 +616,20 @@ KEXT_STATIC int HandleVnodeOperation(
                 SizeOrError xattrResult = Vnode_ReadXattr(currentVnode, PrjFSFileXAttrName, &rootXattr, sizeof(rootXattr));
                 if (xattrResult.error == ENOATTR)
                 {
-                    // Only notify if the file is still a placeholder
+                    // Only notify or block writes if the file is still a placeholder
                     goto CleanupAndReturn;
                 }
                 
+                if (providerStatus == Provider_IsOffline)
```

Comment:
I'm finding the logic in this block a bit hard to follow.  Perhaps some comments would help?

If shouldMessageProvider is false, but providerStatus is Online or Unknown and kauthResult != deny we continue
Then is providerStatus is Offline, we assert if shouldMessageProvider?  

Why does this not follow the normal if (!shouldProvider) goto CleanupAndReturn flow?

Reply:
Walking through the logic top to bottom:

* If the operation would cause no message to be sent to a provider, we're done with the logic, because there's nothing for an offline provider to miss.
* If `TryGetVirtualizationRoot` decided internally that the operation should be blocked, there's also not really anything else to do. I guess the `providerStatus != Provider_IsOffline` check looks redundant in this case, but I'll double-check that before I remove it.

Next, the attribute check:

* If the file is *not* a placeholder (it's a "full" file), we:
 - don't need to `NotifyFilePreConvertToFull`
 - we can allow writes even if the provider is offline because it does no harm (there's no message an offline provider would miss)
* If it *is* a placeholder, we need to send a message if the provider is online, or block the write if it's offline.

I agree a few more comments would help here.

Reply:
Looking at this again, I was right about the `providerStatus != Provider_IsOffline` check.

`shouldMessageProvider` can be `false` if:

 1. The file is not in a virtualisation root.
 2. The provider is offline
 3. It is the provider itself that is causing the I/O.
 4. The I/O is caused by a non-user-initiated process.

Cases 1, 3, & 4 require no further processing.

In the offline provider case, we still need to decide if we need to block the access in case it would lead to a missed message (it would cause a placeholder -> full transition).

I'm restructuring this code slightly, and adding a lot more comments.